pkg/btree/btree.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package btree
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+
+	"github.com/ChainSafe/gossamer/pkg/scale"
+
+	"golang.org/x/exp/constraints"
+
+	"github.com/tidwall/btree"
+)
+
+type Codec interface {
+	MarshalSCALE() ([]byte, error)
+	UnmarshalSCALE(reader io.Reader) error
+}
+
+// Tree is a wrapper around tidwall/btree.BTree that also stores the comparator function and the type of the items
+// stored in the BTree. This is needed during decoding because the Tree item is a generic type, and we need to know it
+// at the time of decoding.
+type Tree struct {
+	*btree.BTree
+	Comparator func(a, b interface{}) bool
```

Comment:
Not sure why you need this attribute given it's not used in any of the `MarshalSCALE` or `UnmarshalSCALE` methods.

Reply:
Please find the follow up questions below to get a better understanding.

Do you mean the comparator? Is the comment not descriptive enough?